src/tonations_generation.py

```diff
@@ -0,0 +1,203 @@
+    tonations_and_points = []
+    for kind in ['major', 'minor']:
+        for base_key in range(12):
+            scale = tonation_to_scale(base_key, kind)
+            tonic = tonation_to_tonic(base_key, kind)
+            subdominant = tonation_to_subdominant(base_key, kind)
+            dominant = tonation_to_dominant(base_key, kind)
+            chord_types = get_aligned_chord_types(
+                sounds, tonic, subdominant, dominant)
+            # chord_types = remove_solitary_chords(chord_types)
+            chord_types = join_same_consecutive_chords(chord_types)
+            points_from_scale = get_points_from_scale_fit(sounds, scale)
+            points_from_chord_patterns = get_points_from_chord_patterns_fit(
+                chord_types)
+            points_from_tail = get_extra_points_for_tail(chord_types)
+            tonation = Tonation(base_key, 0.0, sounds[-1].timestamp, kind)
+            tonations_and_points.append(TonationWithPoints(
+                tonation, points_from_scale,
+                points_from_chord_patterns,
+                points_from_tail
+            ))
+    return tonations_and_points
+
+
+def get_aligned_chord_types(sounds, tonic, subdominant, dominant):
+    tend_to_tonic = True  # handles first-in-tonic = last-in-subdominant uncertainty
+    chord_types = []
+    for sound in sounds:
+        chord_type = sound_to_chord_type(
+            sound, tonic, subdominant, dominant, tend_to_tonic)
+        tend_to_tonic = chord_type != ChordType.SUBDOMINANT
+        chord_types.append(chord_type)
+    return chord_types
+
+
+def sound_to_chord_type(sound, tonic, subdominant, dominant, tend_to_tonic):
+    if sound.note in dominant:
+        return ChordType.DOMINANT
+    if sound.note in tonic:
+        if tend_to_tonic or sound.note not in subdominant:
+            return ChordType.TONIC
+        return ChordType.SUBDOMINANT
+    if sound.note in subdominant:
+        return ChordType.SUBDOMINANT
+    return ChordType.OTHER
+
+
+def remove_solitary_chords(chord_types):
+    if len(chord_types) < 2:
+        return chord_types
+    result = []
+    if chord_types[0] == chord_types[1]:
+        result.append(chord_types[0])
+    for i in range(1, len(chord_types)-1):
+        chord = chord_types[i]
+        if chord_types[i-1] == chord or chord == chord_types[i+1]:
+            result.append(chord)
+    if chord_types[-1] == chord_types[-2]:
+        result.append(chord_types[-1])
+    return result
+
+
+def join_same_consecutive_chords(chord_types):
+    if len(chord_types) < 2:
+        return chord_types
+    result = []
+    last_chord = chord_types[0]
+    result.append(last_chord)
+    for chord in chord_types:
+        if chord != last_chord:
+            last_chord = chord
+            result.append(chord)
+    return result
+
+
+def get_points_from_scale_fit(sounds, scale):
+    points = 0
+    for sound in sounds:
+        if sound.note in scale:
+            points += 10
```

Comment:
Shouldn't it take into consideration duration of sound?

Reply:
I generally agree, but I think decisions like this should be postponed until we have a test dataset